network/p2p/peerstore/peerstore.go

```diff
@@ -249,7 +261,7 @@ func (ps *PeerStore) ReplacePeerList(addressesThey []*peer.AddrInfo, networkName
 
 // AddPersistentPeers stores addresses of peers which are persistent.
 // i.e. they won't be replaced by ReplacePeerList calls
-func (ps *PeerStore) AddPersistentPeers(addrInfo []*peer.AddrInfo, networkName string, role phonebook.PhoneBookEntryRoles) {
+func (ps *PeerStore) AddPersistentPeers(addrInfo []*peer.AddrInfo, networkName string, role phonebook.Role) {
```

Comment:
This is tricky now that we have multiple roles, but a single persistent bit.  If an address is marked to have role A, but not persistent, and then this function used to add role B, I think it will not do so. Lines 268-271 just ensure that the persistent bit is set.

Meanwhile, another bug is lurking.  If this is fixed to set role B and mark the address persistent, it will now have the A role in a persistent way too.

Reply:
Added new checks to prevent persisted peers from modification. AddPersistentPeers updates role if a peer already in a phonebook/peerstore.
Added tests

Reply:
Okay, the previous implementation caused multiple unit tests failure that tested small networks with bootstrap nodes. Had to take another route and make per-role persistence.

Reply:
I like one role at a time, easier to reason about.